samples/bluetooth/hci_uart/src/main.c

```diff
@@ -49,78 +55,136 @@ static K_FIFO_DEFINE(tx_queue);
  */
 #define H4_DISCARD_LEN 33
 
-static int h4_read(struct device *uart, uint8_t *buf,
-		   size_t len, size_t min)
-{
-	int total = 0;
-
-	while (len) {
-		int rx;
+/* Union used for variable into which packet header is received. */
+union bt_hci_hdr {
+	struct bt_hci_cmd_hdr cmd;
+	struct bt_hci_acl_hdr acl;
+	uint8_t raw[0];
+};
 
-		rx = uart_fifo_read(uart, buf, len);
-		if (rx == 0) {
-			LOG_DBG("Got zero bytes from UART");
-			if (total < min) {
-				continue;
-			}
-			break;
-		}
+static int h4_read(struct device *uart, uint8_t *buf, size_t len)
+{
+	int rx = uart_fifo_read(uart, buf, len);
 
-		LOG_DBG("read %d remaining %d", rx, len - rx);
-		len -= rx;
-		total += rx;
-		buf += rx;
-	}
+	LOG_DBG("read %d req %d", rx, len);
 
-	return total;
+	return rx;
 }
 
-static size_t h4_discard(struct device *uart, size_t len)
+static bool valid_type(uint8_t type)
 {
-	uint8_t buf[H4_DISCARD_LEN];
-
-	return uart_fifo_read(uart, buf, MIN(len, sizeof(buf)));
+	return (type == H4_CMD) | (type == H4_ACL);
 }
 
-static void h4_cmd_recv(struct net_buf *buf, int *remaining)
+/* Function assumes that type is validated and only CMD or ACL will be used. */
+static uint32_t get_len(union bt_hci_hdr *hdr, uint8_t type)
 {
-	struct bt_hci_cmd_hdr hdr;
-
-	/* We can ignore the return value since we pass len == min */
-	h4_read(hci_uart_dev, (void *)&hdr, sizeof(hdr), sizeof(hdr));
-
-	*remaining = hdr.param_len;
-
-	net_buf_add_mem(buf, &hdr, sizeof(hdr));
+	return (type == BT_BUF_CMD) ?
+		hdr->cmd.param_len : sys_le16_to_cpu(hdr->acl.len);
+}
 
-	LOG_DBG("len %u", hdr.param_len);
+/* Function assumes that type is validated and only CMD or ACL will be used. */
+static int hdr_len(uint8_t type)
+{
+	return (type == H4_CMD) ?
+		sizeof(struct bt_hci_cmd_hdr) : sizeof(struct bt_hci_acl_hdr);
 }
 
-static void h4_acl_recv(struct net_buf *buf, int *remaining)
+static void rx_isr(void)
 {
-	struct bt_hci_acl_hdr hdr;
+	static struct net_buf *buf;
+	static int remaining;
+	static uint8_t state;
+	static uint8_t type;
+	static union bt_hci_hdr hdr;
+	int read;
+
+	do {
+		switch (state) {
+		case ST_IDLE:
+			/* Get packet type */
+			read = h4_read(hci_uart_dev, &type, sizeof(type));
+			if (read) {
+				if (valid_type(type)) {
+					/* Get expected header size and switch
+					 * to receiving header.
+					 */
+					remaining = hdr_len(type);
+					state = ST_HDR;
+				} else {
+					LOG_WRN("Unknown header %d", state);
+				}
+			}
+			break;
+		case ST_HDR:
+			read = h4_read(hci_uart_dev,
+				       &hdr.raw[hdr_len(type) - remaining],
+				       remaining);
+			remaining -= read;
+			if (remaining == 0) {
+				/* Header received. Get payload length and
+				 * allocate buffer. If allocation fails, discard
+				 * packet.
+				 */
+				remaining = get_len(&hdr, type);
+				buf = bt_buf_get_tx(BT_BUF_H4, K_NO_WAIT,
+						&type, sizeof(type));
+				if (buf == NULL) {
+					return;
```

Comment:
`LOG_WRN` ?

Reply:
this actually could be improved here. Now it will leave the interrupt and hit it again when new data is pending. This data won't be read (because `remaining==0`) so it will keep returning to interrupt. It is the same behavior as in legacy implementation that basically we hit deadlock when there is no buffer.

One way that we could handle that would be to disable RX, start k_timer and in timeout reenable RX, if buffer still cannot be allocated start timer again until buffer can be allocated. Question is if  I should fix it here or maybe it deserves another PR since current patch maintains legacy behavior.

Reply:
`remaining` is not 0 at this point, see [line 126](https://github.com/zephyrproject-rtos/zephyr/pull/27918/files#diff-4518902872ac572c659d96216b345521R126).

Reply:
right, i would swap setting remaininig with getting buf to ensure that it will stay in the same state (ST_HDR, remaning = 0).

Reply:
Update also the comment above that line. It mentions discarding the packet (and that's why I asked https://github.com/zephyrproject-rtos/zephyr/pull/27918#discussion_r481115418).

Reply:
updated the comment and added assert because currently (and previously) this case is not supported. I started to look into somehow handling it but it is not that trivial so i don't want that in this PR.